README.md

```diff
@@ -1,486 +1,679 @@
+- A decorated setter is called with the setter function being decorated
+
+(Decorators are called with a second parameter giving more context, but we don't need those details for the `@logged` decorator.)
+
+The return value of a decorator is a new value that replaces the thing it's wrapping. For methods, getters and setters, the return value is another function to replace that method, getter or setter.
 
 ```mjs
 // logged.mjs
 
-export decorator @logged {
-  @wrap(f => {
-    const name = f.name;
-    function wrapped(...args) {
-      console.log(`starting ${name} with arguments ${args.join(", ")}`);
-      f.call(this, ...args);
-      console.log(`ending ${name}`);
-    }
-    wrapped.name = name;
-    return wrapped;
-  })
+export function logged(f) {
+  const name = f.name;
+  function wrapped(...args) {
+    console.log(`starting ${name} with arguments ${args.join(", ")}`);
+    const ret = f.call(this, ...args);
+    console.log(`ending ${name}`);
+    return ret;
+  }
+  wrapped.name = name;
+  return wrapped;
 }
 ```
 
-In the above example, the *composed decorator* `@logged` is defined to expand out into a call of the `@wrap` decorator with a particular fixed callback.
+This example roughly "desugars" to the following (i.e., could be transpiled as such):
 
-### `@defineElement`
+```js
+let x_setter;
+
+class C {
+  method(arg) {
+    this.#x = arg;
+  }
+
+  static #x_setter(value) { }
+  static #_ = (x_setter = C.#x_setter, void 0);
+  set #x(value) { return x_setter.call(this, value); }
+}
+
+C.prototype.method = logged(C.prototype.method, { kind: "method", name: "method", isStatic: false });
+x_setter = logged(x_setter, {kind: "setter", isStatic: false});
+```
+
+Note that getters and setters are decorated separately. Accessors are not "coalesced" as in earlier decorators proposals  (unless they are generated for a field; see below).
+
+## `@defineElement`
 
 [HTML Custom Elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) lets you define your own HTML element. Elements are registered using `customElements.define`. Using decorators, the registration can happen up-front:
 
 ```mjs
-import { @defineElement } from "./defineElement.mjs";
+import { defineElement } from "./defineElement.mjs";
 
 @defineElement('my-class')
 class MyClass extends HTMLElement { }
 ```
 
-The `@defineElement` decorator is based on the `@register` decorator. This decorator is called when the class is finished being defined, and it calls the provided callback with the class that resulted.
+Classes can be decorated just like methods and accessors. The class shows up in the `value` option.
 
 ```mjs
 // defineElement.mjs
-export decorator @defineElement(name, options) {
-  @register(klass => customElements.define(name, klass, options))
+export function defineElement(name, options) {
+  return klass => { customElements.define(name, klass, options); return klass; }
```

Comment:
presumably omitting the return/returning undefined would be equivalent?

Reply:
We could decide that one way or the other; I was thinking about just requiring that you actually include the return statement.